File: algo/QuickSort/quicksort_basic.go
package main

import "fmt"

func quicksort(arr []int) {
	fmt.Println("init arr = ", arr)
	//quicksort_basic(arr, 0, len(arr)-1)
	//quicksort_twopointers(arr, 0, len(arr)-1)
	quicksort_advance(arr, 0, len(arr) - 1)
}
func quicksort_basic(arr []int, start int, end int) {
	if start >= end {
		return
	}
	middle := partition(arr, start, end)
	quicksort_basic(arr, start, middle-1)
	quicksort_basic(arr, middle+1, end)
}
func partition(arr []int, start int, end int) int {
	pivot := arr[start]
	left := start + 1
	right := end
	for left < right {
		for left < right && arr[left] <= pivot {
			left++
		}
		fmt.Printf("left = %d, right = %d\n", left, right)
		if left != right {
			arr[left], arr[right] = arr[right], arr[left]
			fmt.Println(arr)
			right--
		}
	}
	if left == right && arr[right] > pivot {
		right--
	}
	if right != start {
		arr[start], arr[right] = arr[right], arr[start]
		fmt.Println(arr)
	}
	return right
}

func quicksort_twopointers(arr []int, start int, end int) {
	if start >= end {
		return
	}
	middle := partition_twopointers(arr, start, end)
	quicksort_twopointers(arr, start, middle-1)
	quicksort_twopointers(arr, middle+1, end)
}
func partition_twopointers(arr []int, start int, end int) int {
	pivot := arr[start]
	left := start + 1
	right := end
	for left < right {
		for left < right && arr[left] <= pivot {
			left++
		}
		for left < right && arr[right] >= pivot {
			right--
		}
		fmt.Printf("left = %d, right = %d\n", left, right)
		if left != right {
			arr[left], arr[right] = arr[right], arr[left]
			fmt.Println(arr)
			left++
			right--
		}
	}
	if left == right && arr[right] > pivot {
		right--
	}
	if right != start {
		arr[start], arr[right] = arr[right], arr[start]
		fmt.Println(arr)
	}
	return right
}
func main() {
	arr := []int{4, 2, 7, 1, 6, 3, 5}
	quicksort(arr)
}

/*
代碼出處:
https://leetcode.cn/leetbook/read/sort-algorithms/eul7hm/

執行指令
go run ./algo/QuickSort
*/

File: algo/QuickSort/quickSort_basic.py
class quickSort:
    @classmethod
    def quickSort(cls, arr):
        #選擇一種quickSort執行
        #cls.quickSort_basic(arr, 0, len(arr) - 1)#注意, 這邊要-1
        cls.quickSort_twopointers(arr, 0, len(arr) - 1)

    @classmethod
    def quickSort_basic(cls, arr, start, end):
        if start >= end:
            return
        middle = cls.partition(arr, start, end)#注意, middle是這樣來的
        cls.quickSort_basic(arr, start, middle - 1)
        cls.quickSort_basic(arr, middle + 1, end)

    @classmethod
    def partition(cls, arr, start, end):
        pivot = arr[start]
        left = start + 1
        right = end
        while left < right:
            while left < right and pivot > arr[left]:
                left += 1
            if left != right:
                arr[left], arr[right] = arr[right], arr[left]
                right -= 1
            print("After swap", arr)

        if left == right and arr[right] > pivot:
            right -= 1
        arr[start], arr[right] = arr[right], arr[start]
        return right

    @classmethod
    def quickSort_twopointers(cls, arr, start, end):
        if start >= end:
            return
        middle = cls.partition_twopointers(arr, start, end)
        cls.quickSort_twopointers(arr, start, middle - 1)
        cls.quickSort_twopointers(arr, middle + 1, end)

    @classmethod
    def partition_twopointers(cls, arr, start, end):
        pivot = arr[start]
        left = start + 1
        right = end
        while left < right:
            while left < right and arr[left] < pivot:
                left += 1
            while left < right and arr[right] > pivot:
                right -= 1
            if left < right:
                arr[left], arr[right] = arr[right], arr[left]
                left += 1
                right -= 1
                
        #注意, 在推倒時不要忘了這一句
        if left == right and arr[right] > pivot:
            right -= 1
        arr[start], arr[right] = arr[right], arr[start]
        return right
        
if __name__ == "__main__":
    arr = [4, 2, 7, 1, 6, 3, 5]
    # quickSort.partition_twopointers(arr, 0, len(arr) - 1)
    quickSort.quickSort(arr)
    print(arr)
